### PR TITLE
Add measure tool

### PR DESCRIPTION
**Pull request by kandre**
For #7

### CODE REVIEW

**Comment by kandre** [comment]
Bug on chrome: the tooltip 'on last point' does not refresh, while FF does

**Comment by kandre** [comment]
Bugs are fixed (update of help tooltip messages on last/first point)

It works now on chrome, FF and IE

I saw that div 'Measure' / 'Cancel measure mode' no longer update if a language is changed during a drawing. I am setting that now.

File: js/measure.js [hunk 1]
```diff
@@ -0,0 +1,427 @@
...
+    helpTooltipElement = document.createElement('div');
+    // Name generated element
+    helpTooltipElement.setAttribute('id', 'helpTooltip');
+    helpTooltipElement.className = 'tooltip-help';
+    helpTooltip = new ol.Overlay({
+      element: helpTooltipElement,
+      offset: [15, 0],
+      positioning: 'center-left'
+    });
+    map.addOverlay(helpTooltip);
```
Comment (oterral) [inline js/measure.js]: I think you shouldn't add the overlay here but in the activate method and remove it in deactivate method. Same for all tooltip elements.

File: js/measure.js [hunk 2]
```diff
@@ -0,0 +1,427 @@
...
+    createAreaTooltip();
+    createDistanceTooltip();
+    createHelpTooltip();
+  };
+
+  var deactivate = function() {
+    ol.Observable.unByKey(deregKey);
+    ol.Observable.unByKey(deregKey2);
+    drawPolygon.setActive(false);
+    layer.getSource().clear();
```
Comment (oterral) [inline js/measure.js]: No need to clear the source keep the elements drawn. When the user reactivates the measure, he wants to see what he has drawn I guess

File: js/measure.js [hunk 3]
```diff
@@ -0,0 +1,427 @@
...
+    map.removeLayer(layer);
+    //helpTooltipElement.parentNode.removeChild(helpTooltipElement);
+    $('#helpTooltip').remove();
+    $('#area' + (areaTooltipId - 1).toString()).remove();
+    $('#distance' + (distanceTooltipId - 1).toString()).remove();
```
Comment (oterral) [inline js/measure.js]: Remove the overlay instead of removing elements

File: js/measure.js [hunk 4]
```diff
@@ -0,0 +1,427 @@
...
+    }
+  });
+  drawPolygon.setActive(false);
+  drawPolygon.on('drawstart', function(evt) {
+
+    var nbPoint = 1;
+    isSnapOnLastPoint = false;
+
+    // Clear the layer
+    layer.getSource().clear();
```
Comment (oterral) [inline js/measure.js]: You must also remove the overlays you add to display the length or area

File: js/measure.js [hunk 5]
```diff
@@ -0,0 +1,427 @@
...
+
+    // Clear the previous measure tooltips
+    if (areaTooltipId > 1) {
+      $('#area' + (areaTooltipId - 1).toString()).remove();
+    }
+    if (distanceTooltipId > 1) {
+      $('#distance' + (distanceTooltipId - 1).toString()).remove();
+    }
```
Comment (oterral) [inline js/measure.js]: Remove the overlays instead of elements.

File: js/measure.js (hunk 2 above)
Comment (kandre) [inline js/measure.js]: Ok, yes that makes sense

File: js/measure.js [hunk 6]
```diff
@@ -0,0 +1,426 @@
+var addMeasure = function(map, translator) {
+
+  // Currently drawn feature.
+  var sketch;
```
Comment (oterral) [inline js/measure.js]: skecth seems to be used only once, you can remove it I guess

File: js/measure.js [hunk 7]
```diff
@@ -0,0 +1,426 @@
...
+      var lineCoords = sketchFeatDistance.getGeometry().getCoordinates();
+      lineCoords.pop();
+      sketchFeatDistance.getGeometry().setCoordinates(lineCoords);
+    }          
+
+    // Update the layer
+    updateLayer(isFinishOnFirstPoint);
+
+    // Unregister the change event
+    sketchFeatArea.unByKey(deregisterFeature);
```
Comment (oterral) [inline js/measure.js]: use `ol.Observable.unByKey`

File: js/measure.js [hunk 8]
```diff
@@ -0,0 +1,426 @@
...
+      element: helpTooltipElement,
+      offset: [15, 0],
+      positioning: 'center-left'
+    });
+  };
+
+  // Creates a new measure area tooltip
+  var createAreaTooltip = function() {
+    var areaTooltipElement = document.createElement('div');
+    areaTooltipElement.setAttribute('id', 'area');
```
Comment (oterral) [inline js/measure.js]: Remove the definition of an id if it's not used. 
Same for other tooltip

File: js/measure.js [hunk 9]
```diff
@@ -0,0 +1,426 @@
...
+      if (isSnapOnLastPoint === true) {
+        helpMsg = translator.get('measureMsgOnLastPoint');
+      } else if (!isFinishOnFirstPoint) {
+        helpMsg = continueDrawingMsg;
+      }
+
+      if (geom instanceof ol.geom.Polygon) {
+        output = formatArea(geom);
+        tooltipCoord = geom.getInteriorPoint().getCoordinates();
+        areaTooltipCurrent.get('element').innerHTML = output;
```
Comment (oterral) [inline js/measure.js]: use `getElement`function

File: js/measure.js [hunk 10]
```diff
@@ -0,0 +1,426 @@
...
+          areaTooltipCurrent.setPosition(tooltipCoord);
+        } else {
+          // Position is hidden if pointer not on first point
+          areaTooltipCurrent.setPosition(undefined);
+        }
+
+      } else if (geom instanceof ol.geom.LineString) {
+        output = formatLength(geom);
+        tooltipCoord = geom.getLastCoordinate();
+        distanceTooltipCurrent.get('element').innerHTML = output;
```
Comment (oterral) [inline js/measure.js]: use getElementfunction

File: js/measure.js [hunk 11]
```diff
@@ -0,0 +1,426 @@
...
+      } else if (!isSnapOnLastPoint) {
+        helpMsg = continueDrawingMsg;
+      }
+
+      // Condition for changing help tooltip if pointer snap to first point
+      if (isSnapOnLastPoint === true) {
+        helpMsg = translator.get('measureMsgOnLastPoint');
+      } else if (!isFinishOnFirstPoint) {
+        helpMsg = continueDrawingMsg;
+      }
```
Comment (oterral) [inline js/measure.js]: Something is weird why the `helpMessage` stuff is    here ? I think it should be in the point move handler

File: js/measure.js [hunk 12]
```diff
@@ -0,0 +1,426 @@
...
+    var output;
+    if (area > 10000) {
+      output = (Math.round(area / 1000000 * 100) / 100) + ' ' + 'km<sup>2</sup>';
+    } else {
+      output = (Math.round(area * 100) / 100) + ' ' + 'm<sup>2</sup>';
+    }
+    return output;
+  };
+ 
+  var renderTooltip = function(sketchFeatArea, sketchFeatDistance) {
```
Comment (oterral) [inline js/measure.js]: Add comment to this function otherwise we are not sure which tooltip should be rendered

File: js/measure.js [hunk 13]
```diff
@@ -0,0 +1,426 @@
...
+      }
+    );
+  });
+
+  drawPolygon.on('drawend', function(evt) {
+    // Display static tooltip
+    if (isFinishOnFirstPoint) {
+      // Clone current tooltip to generate a static tooltip
+      areaTooltipStatic = areaTooltipCurrent;
+      areaTooltipStatic.get('element').className = 'tooltip tooltip-static';
```
Comment (oterral) [inline js/measure.js]: `getElement`

File: js/measure.js [hunk 14]
```diff
@@ -0,0 +1,426 @@
...
+      areaTooltipStatic.setOffset([0, -7]);
+      map.addOverlay(areaTooltipStatic);
+      map.removeOverlay(distanceTooltipCurrent);
+    } else {
+      // Clone current tooltip to generate static tooltip
+      distanceTooltipStatic = distanceTooltipCurrent;
+      distanceTooltipStatic.get('element').className = 'tooltip tooltip-static';
```
Comment (oterral) [inline js/measure.js]: `getElement`

File: js/measure.js (hunk 8 above)
Comment (kandre) [inline js/measure.js]: yes I only use classes in the css

Comment (kandre) [inline js/measure.js]: However I use an id for help tooltip, which seems convenient to pass messages

File: js/measure.js [hunk 15]
```diff
@@ -0,0 +1,426 @@
...
+
+  map.addInteraction(drawPolygon);
+  
+  var deregKey, deregKey2;
+  var activate = function() {
+    drawPolygon.setActive(true);
+    deregKey = map.on('pointermove', pointerMoveHandler);
+    deregKey2 = map.on('click', function(evt) {
+      // Refresh help tooltip after a click (not working on chrome)
+      $('#helpTooltip').text(translator.get('measureMsgOnLastPoint'));
```
Comment (oterral) [inline js/measure.js]: you have the element,  you can use it : `$(helpElement).text`

Comment (kandre) [inline js/measure.js]: definitely ! with a getElement as well

File: js/measure.js (hunk 11 above)
Comment (kandre) [inline js/measure.js]: This render function returns the helpMsg according to events such as snap on first or last point, therefore I thought it is the proper place.

**Comment by kandre** [comment]
@oterral ,
- In my last commit I removed an useless (in my opinion) call of the updateHelpLabel function at the beginning of the 'drawstart' event.
- I initialize the help message in the activate function

I'd say now it is bug-free

File: js/measure.js [hunk 16]
```diff
@@ -0,0 +1,418 @@
...
+    if (area > 10000) {
+      output = (Math.round(area / 1000000 * 100) / 100) + ' ' + 'km<sup>2</sup>';
+    } else {
+      output = (Math.round(area * 100) / 100) + ' ' + 'm<sup>2</sup>';
+    }
+    return output;
+  };
+ 
+  // Rendering help tooltip
+  var renderTooltip = function(sketchFeatArea, sketchFeatDistance) {
```
Comment (oterral) [inline js/measure.js]: that renders all but tnot the help tooltip

File: js/measure.js [hunk 17]
```diff
@@ -0,0 +1,418 @@
...
+    }
+    isSnapOnLastPoint = false;
+     
+    // Update help label
+    updateHelpLabel();
+  });
+
+  map.addInteraction(drawPolygon);
+  
+  var deregKey, deregKey2;
```
Comment (oterral) [inline js/measure.js]: remove deregKey2

**Comment by oterral** [comment]
2 minor comments, fixes it and squash !!!  awesome work @kandre !!

**Comment by oterral** [comment]
Found other things: 
  - add `z-index:900` to the map , add `z-index:1000` to the catalog
  - Fix translation of measure link and help start tooltip

**Comment by kandre** [comment]
Pushed !

File: js/measure.js [hunk 18]
```diff
@@ -0,0 +1,428 @@
...
+    map.addOverlay(distanceTooltipCurrent);
+    // Reclaim distance measure tooltip if already been drawn
+    if (distanceTooltipStatic) {
+      map.addOverlay(distanceTooltipStatic);
+    }
+
+    createHelpTooltip();
+    map.addOverlay(helpTooltip);
+    var helpMsg = translator.get('measureMsgStart');
+    $('.tooltip-help').attr('data-value', 'measureMsgStart');
```
Comment (oterral) [inline js/measure.js]: use `$(tooltipElement)`everey where

File: js/translation.js [hunk 19]
```diff
@@ -6,6 +6,13 @@ var addTranslation = function(objPermalink, translator) {
     $('#logo img').attr('src', translator.get('logo'));
     $('.linkcopyright').attr('href', 'http://www.geo.admin.ch/internet/geoportal/' + lang + '/home/geoadmin/contact.html#copyright');
     $('.ol-attribution a').attr('href', 'http://www.swisstopo.admin.ch/internet/swisstopo/' + lang + '/home.html');
+    if ($('#measure').attr('active') == 1) {
+      $('#measure').text(translator.get('cancelMeasure'));
+    } else {
+      $('#measure').text(translator.get('measure'));
```
Comment (oterral) [inline js/translation.js]: var measure = $('#measure');
measaure.text(translator.get(measaure.attr('data-value')));

**Comment by kandre** [comment]
Squashed

**Comment by kandre** [comment]
Thanks !!

**Pull request merged**